Replace encoder with url encoder

salvatore-campagna opened the pull request:
Document IDs are frequently used in HTTP requests, such as `GET /index/_doc/{id}`, where they must be URL-safe to avoid issues with invalid characters. This change ensures that IDs generated by `TimeBasedKOrderedUUIDGenerator` are properly Base64 URL-encoded, free of characters that could break URLs. We also test that no IDs include invalid characters like +, /, or = to guarantee they are fully compliant with URL-safe requirements.

Moreover `TimeBasedKOrderedUUIDGenerator` and `TimeBasedUUIDGenerator` are refactored to allow injection of dependencies which enables us to increase test coverage by including tests for high-throughput scenarios, sequence id overflow and unreliable clocks usage.

salvatore-campagna committed:
fix: replace encoder with url encoder

Dcoument IDs might be used in requests like get by id
which might include invalid characters if the id is not
url encoded.

Here we also test that, no id is generated including
possible invalid characters.

salvatore-campagna committed:
test: try also to decode a url-encoded id

salvatore-campagna commented:
Testing this with the previous encoder actually resulted in failures

```
UUIDTests > testKOrderedUUID FAILED
    java.lang.AssertionError: UUID contains invalid characters: AZMi5TapRhFlU+F2NPIJ
        at __randomizedtesting.SeedInfo.seed([34991382834991A7:16CA98F0E1984993]:0)
        at org.junit.Assert.fail(Assert.java:89)
        at org.junit.Assert.assertTrue(Assert.java:42)
        at org.elasticsearch.common.UUIDTests.verifyUUIDIsUrlSafe(UUIDTests.java:245)
        at java.base/java.lang.Iterable.forEach(Iterable.java:75)
        at org.elasticsearch.common.UUIDTests.testKOrderedUUID(UUIDTests.java:51)

REPRODUCE WITH: ./gradlew ":server:test" --tests "org.elasticsearch.common.UUIDTests.testThreadedKOrderedUUID" -Dtests.seed=34991382834991A7 -Dtests.locale=ja-Jpan-JP -Dtests.timezone=America/Belize -Druntime.java=23

UUIDTests > testThreadedKOrderedUUID FAILED
    java.lang.AssertionError: UUID contains invalid characters: AZMi5TapRhFlU+OBp/Xa
        at __randomizedtesting.SeedInfo.seed([34991382834991A7:F891DF8432167CEE]:0)
        at org.junit.Assert.fail(Assert.java:89)
        at org.junit.Assert.assertTrue(Assert.java:42)
        at org.elasticsearch.common.UUIDTests.verifyUUIDIsUrlSafe(UUIDTests.java:245)
        at java.base/java.lang.Iterable.forEach(Iterable.java:75)
        at org.elasticsearch.common.UUIDTests.testUUIDThreaded(UUIDTests.java:148)
        at org.elasticsearch.common.UUIDTests.testThreadedKOrderedUUID(UUIDTests.java:63)
 ```

salvatore-campagna committed:
test: just try to decode the id

salvatore-campagna committed:
test: do not check the charset

kkrik-es commented on server/src/main/java/org/elasticsearch/common/TimeBasedKOrderedUUIDGenerator.java:
```diff
@@ -68,6 +68,6 @@ public String getBase64UUID() {
         assert buffer.position() == uuidBytes.length;
 
-        return BASE_64_NO_PADDING.encodeToString(uuidBytes);
+        return BASE_64_URL_ENCODER_NO_PADDING.encodeToString(uuidBytes);
```
Let's just use `Strings.BASE_64_NO_PADDING_URL_ENCODER`, it's the same and used elsewhere too.

kkrik-es commented on server/src/test/java/org/elasticsearch/common/UUIDTests.java:
```diff
@@ -237,4 +240,13 @@ public void testStringLength() {
+    private void verifyUUIDIsUrlSafe(final String uuid) {
+        assertFalse("UUID should not contain padding characters: " + uuid, uuid.contains("="));
+        try {
+            BASE_64_URL_DECODER.decode(uuid);
```
I may've missed it, but it'd be nice to test that encoding and decoding returns the original byte array, for random inputs.

martijnvg replied:
+1 to such a test

kkrik-es reviewed (approved):
Thanks Salvatore. A few nits, nothing major.

martijnvg reviewed (approved):
LGTM

salvatore-campagna committed:
fix: use existing url encoder

salvatore-campagna committed:
refactor: make TimeBasedUUIDGenerator and TimeBasedKOrderedUUIDGenerator testable

Here we use dependency injection to make both TimeBasedUUIDGenerator and
TimeBasedKOrderedUUIDGenerator testable using unit tests. This way we
can test edge cases like high-throughput document ingestion, unreliable
clocks and sequence ID overflow.

salvatore-campagna commented:
@elasticmachine update branch

elasticmachine committed:
Merge branch 'main' into fix/doc-id-url-encode

salvatore-campagna committed:
test: add a few more test and refactor existing ones

salvatore-campagna commented:
@elasticmachine update branch

elasticmachine committed:
Merge branch 'main' into fix/doc-id-url-encode

salvatore-campagna committed:
nit: remove count variable

salvatore-campagna committed:
fix: remove unused decoder

salvatore-campagna committed:
nit: re-arrange some code and remove unused constant

salvatore-campagna commented on server/src/main/java/org/elasticsearch/common/TimeBasedKOrderedUUIDGenerator.java:
```diff
@@ -28,18 +28,25 @@
 public class TimeBasedKOrderedUUIDGenerator extends TimeBasedUUIDGenerator {
-    private static final Base64.Encoder BASE_64_NO_PADDING = Base64.getEncoder().withoutPadding();
+
+    public TimeBasedKOrderedUUIDGenerator(
```
I did this change here to be able to use dependency injection in unit tests.

salvatore-campagna commented on server/src/test/java/org/elasticsearch/common/TimeBasedUUIDGeneratorTests.java:
```diff
@@ -0,0 +1,270 @@
+import java.util.function.Supplier;
+import java.util.stream.IntStream;
+
+public class TimeBasedUUIDGeneratorTests extends ESTestCase {
```
I have specific unit tests for both `TimeBasedUUIDGenerator` and `TimeBasedKOrderedUUIDGenerator`...both of them generate UUID based on time and share the same requirements...indeed the only difference between them is the way bytes are ordered.

salvatore-campagna commented:
@elasticmachine update branch

elasticmachine committed:
Merge branch 'main' into fix/doc-id-url-encode

elasticsearchmachine commented:
## 💔 Backport failed
The backport operation could not be completed due to the following error:
```
An unexpected error occurred when attempting to backport this PR.
```

You can use [sqren/backport](https://github.com/sqren/backport) to manually backport by running `backport --upstream elastic/elasticsearch --pr 116699`

salvatore-campagna commented:
## 💚 All backports created successfully

| Status | Branch | Result |
|:------:|:------:|:------|
|✅|8.x|[<img src="https://img.shields.io/github/pulls/detail/state/elastic/elasticsearch/119079">](https://github.com/elastic/elasticsearch/pull/119079)|
|✅|8.17|[<img src="https://img.shields.io/github/pulls/detail/state/elastic/elasticsearch/119080">](https://github.com/elastic/elasticsearch/pull/119080)|


### Questions ?
Please refer to the [Backport tool documentation](https://github.com/sqren/backport)
